// @ramp-network/ramp-instant-sdk/dist/lib/ramp-instant-sdk.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var body_scroll_lock_1 = require("body-scroll-lock");
var consts_1 = require("./consts");
var event_polling_1 = require("./event-polling");
var init_helpers_1 = require("./init-helpers");
var types_1 = require("./types");
var utils_1 = require("./utils");
var types_2 = require("./types");
exports.RampInstantEventTypes = types_2.WidgetEventTypes;
var RampInstantSDK = /** @class */ (function () {
    function RampInstantSDK(config) {
        this._listeners = utils_1.initEventListenersDict();
        this._isVisible = false;
        this._isPollingForSwapStatus = false;
        this._purchasePollingCredentials = null;
        init_helpers_1.importFonts();
        this.unsubscribe = this.unsubscribe.bind(this);
        this.on = this.on.bind(this);
        this.show = this.show.bind(this);
        this._handleEscapeClick = this._handleEscapeClick.bind(this);
        this._dispatchEvent = this._dispatchEvent.bind(this);
        this._subscribeToWidgetEvents = this._subscribeToWidgetEvents.bind(this);
        this._on = this._on.bind(this);
        this._processPurchasePollingLoop = this._processPurchasePollingLoop.bind(this);
        this._registerSdkEventHandlers = this._registerSdkEventHandlers.bind(this);
        this._runPostSubscribeHooks = this._runPostSubscribeHooks.bind(this);
        this._subscribeToWidgetEvents = this._subscribeToWidgetEvents.bind(this);
        var temporaryConfig = utils_1.normalizeConfigAndLogErrorsOnInvalidFields(__assign({ variant: 'desktop' }, config));
        var widgetVariant = utils_1.determineWidgetVariant(temporaryConfig);
        this._config = __assign(__assign({}, temporaryConfig), { variant: widgetVariant, widgetInstanceId: utils_1.getRandomIntString() });
    }
    RampInstantSDK.prototype.show = function () {
        var _a, _b;
        if (this._isVisible) {
            throw new Error('Widget is already visible - you can only call this once per instance');
        }
        if (document.activeElement && utils_1.isHtmlElement(document.activeElement)) {
            document.activeElement.blur();
        }
        var widgetUrl = init_helpers_1.initWidgetIframeUrl(this._config);
        this._registerSdkEventHandlers();
        window.addEventListener('message', this._subscribeToWidgetEvents);
        if (this._isConfiguredAsHosted()) {
            this.widgetWindow = (_a = window.open(widgetUrl), (_a !== null && _a !== void 0 ? _a : undefined));
            return this;
        }
        this.domNodes = init_helpers_1.initDOMNode(widgetUrl, this._dispatchEvent, this._config);
        if (!((_b = this.domNodes) === null || _b === void 0 ? void 0 : _b.body)) {
            throw new Error("Couldn't find <body> element.");
        }
        this.domNodes.body.appendChild(this.domNodes.shadowHost);
        this._isVisible = true;
        body_scroll_lock_1.default.disableBodyScroll(this.domNodes.iframe);
        window.addEventListener('keydown', this._handleEscapeClick, true);
        return this;
    };
    RampInstantSDK.prototype.on = function (type, callback) {
        this._on(type, callback, false);
        return this;
    };
    RampInstantSDK.prototype.unsubscribe = function (type, callback) {
        var _this = this;
        if (type === '*') {
            var allTypes = Object.entries(this._listeners);
            allTypes.forEach(function (_a) {
                var key = _a[0], eventHandlers = _a[1];
                var filteredHandlers = eventHandlers.filter(function (l) { return l.callback !== callback; });
                _this._listeners[key] = filteredHandlers;
            });
        }
        else {
            this._listeners[type] = this._listeners[type].filter(function (l) { return l.callback !== callback; });
        }
        return this;
    };
    RampInstantSDK.prototype._on = function (type, callback, internal) {
        if (type !== '*' && !this._listeners[type]) {
            // tslint:disable-next-line:no-console
            console.warn("Unknown / unsupported event name - '" + type + "'. This listener will have no effect.");
        }
        if (type === '*') {
            var allTypes = Object.values(this._listeners);
            allTypes.forEach(function (eventHandlers) { return eventHandlers.push({ callback: callback, internal: internal }); });
        }
        else {
            this._listeners[type].push({ callback: callback, internal: internal });
        }
        this._runPostSubscribeHooks(type, internal);
    };
    RampInstantSDK.prototype._subscribeToWidgetEvents = function (event) {
        if (!event.data) {
            return;
        }
        if (!init_helpers_1.areUrlsEqual(event.origin, this._config.url || consts_1.baseWidgetUrl)) {
            return;
        }
        var eventData = event.data;
        if (!eventData.widgetInstanceId ||
            eventData.widgetInstanceId !== this._config.widgetInstanceId) {
            return;
        }
        this._dispatchEvent(eventData);
    };
    RampInstantSDK.prototype._registerSdkEventHandlers = function () {
        var _this = this;
        this._on(types_1.WidgetEventTypes.WIDGET_CLOSE, function (_event) {
            var _a, _b;
            if (_this._isConfiguredAsHosted()) {
                try {
                    (_a = _this.widgetWindow) === null || _a === void 0 ? void 0 : _a.close();
                }
                catch (e) {
                    throw new Error('Could not close the widget window');
                }
            }
            else {
                (_b = _this.domNodes) === null || _b === void 0 ? void 0 : _b.shadowHost.remove();
                body_scroll_lock_1.default.clearAllBodyScrollLocks();
            }
            _this._teardownEventSubscriptions();
        }, true);
        var onConfigEvent = function () {
            var _a, _b;
            if (_this._isConfiguredAsHosted()) {
                return;
            }
            (_a = _this.domNodes) === null || _a === void 0 ? void 0 : _a.iframe.classList.add('visible');
            var loader = (_b = _this.domNodes) === null || _b === void 0 ? void 0 : _b.overlay.querySelector('.loader-container');
            if (loader) {
                loader.remove();
            }
        };
        this._on(types_1.WidgetEventTypes.WIDGET_CONFIG_DONE, onConfigEvent, true);
        this._on(types_1.WidgetEventTypes.WIDGET_CONFIG_FAILED, onConfigEvent, true);
        this._on(types_1.InternalEventTypes.WIDGET_CLOSE_REQUEST, function (_event) {
            if (_this._isConfiguredAsHosted()) {
                return;
            }
            if (_this._config.variant === 'mobile' || init_helpers_1.isCloseModalAlreadyOpen(_this.domNodes.overlay)) {
                return;
            }
            _this.domNodes.overlay.appendChild(init_helpers_1.prepareCloseModalNode(_this._dispatchEvent));
        }, true);
        this._on(types_1.InternalEventTypes.WIDGET_CLOSE_REQUEST_CONFIRMED, function (_event) {
            _this._dispatchEvent({
                type: types_1.WidgetEventTypes.WIDGET_CLOSE,
                payload: null,
                widgetInstanceId: _this._config.widgetInstanceId,
            });
        }, true);
        this._on(types_1.InternalEventTypes.WIDGET_CLOSE_REQUEST_CANCELLED, function (_event) {
            if (_this._isConfiguredAsHosted()) {
                return;
            }
            var modal = _this.domNodes.overlay.querySelector('.close-modal');
            if (modal) {
                modal.remove();
            }
        }, true);
        this._on(types_1.WidgetEventTypes.PURCHASE_CREATED, function (event) {
            _this._purchasePollingCredentials = {
                apiUrl: event.payload.apiUrl,
                purchaseExternalId: event.payload.purchase.id,
                token: event.payload.purchaseViewToken,
            };
            // tslint:disable-next-line:no-floating-promises
            _this._processPurchasePollingLoop(_this._purchasePollingCredentials);
        }, true);
    };
    RampInstantSDK.prototype._dispatchEvent = function (event) {
        var type = event.type;
        this._listeners[type].forEach(function (handler) { return handler.callback(event); });
    };
    RampInstantSDK.prototype._handleEscapeClick = function (event) {
        var escKeyCode = 27;
        if (event.key === 'Escape' || event.key === 'Esc' || event.keyCode === escKeyCode) {
            this._dispatchEvent({
                type: types_1.InternalEventTypes.WIDGET_CLOSE_REQUEST,
                payload: null,
                internal: true,
            });
        }
    };
    // Event subscriptions aren't cleared so that host can receive a PAYMENT_SUCCESSFUL event
    // even after the widget has been closed
    RampInstantSDK.prototype._teardownEventSubscriptions = function () {
        window.removeEventListener('keydown', this._handleEscapeClick, true);
        window.removeEventListener('message', this._subscribeToWidgetEvents);
    };
    RampInstantSDK.prototype._processPurchasePollingLoop = function (_a) {
        var apiUrl = _a.apiUrl, purchaseExternalId = _a.purchaseExternalId, token = _a.token;
        return __awaiter(this, void 0, void 0, function () {
            var purchase;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        this._isPollingForSwapStatus = true;
                        _b.label = 1;
                    case 1:
                        if (!(utils_1.countListenersForEvent(this._listeners, types_1.WidgetEventTypes.PURCHASE_SUCCESSFUL) > 0 ||
                            utils_1.countListenersForEvent(this._listeners, types_1.WidgetEventTypes.PURCHASE_FAILED) > 0)) return [3 /*break*/, 4];
                        // tslint:disable-next-line:no-magic-numbers
                        return [4 /*yield*/, event_polling_1.delay(1000)];
                    case 2:
                        // tslint:disable-next-line:no-magic-numbers
                        _b.sent();
                        return [4 /*yield*/, event_polling_1.doFetchPurchase(apiUrl, purchaseExternalId, token)];
                    case 3:
                        purchase = _b.sent();
                        if (!purchase) {
                            return [3 /*break*/, 1];
                        }
                        if (purchase.actions.find(function (action) { return action.newStatus === 'RELEASED'; })) {
                            this._dispatchEvent({
                                type: types_1.WidgetEventTypes.PURCHASE_SUCCESSFUL,
                                payload: {
                                    purchase: purchase,
                                },
                                widgetInstanceId: this._config.widgetInstanceId,
                            });
                            return [2 /*return*/];
                        }
                        if (purchase.actions.find(function (action) { return action.newStatus === 'ERROR'; })) {
                            this._dispatchEvent({
                                type: types_1.WidgetEventTypes.PURCHASE_FAILED,
                                payload: null,
                                widgetInstanceId: this._config.widgetInstanceId,
                            });
                            return [2 /*return*/];
                        }
                        return [3 /*break*/, 1];
                    case 4:
                        this._isPollingForSwapStatus = false;
                        return [2 /*return*/];
                }
            });
        });
    };
    RampInstantSDK.prototype._runPostSubscribeHooks = function (eventType, isHandlerInternal) {
        /*
         *  Handles a case where a host subscribes to these events after
         *  the `PURCHASE_CREATED` event is fired
         */
        if ((eventType === '*' ||
            eventType === types_1.WidgetEventTypes.PURCHASE_SUCCESSFUL ||
            types_1.WidgetEventTypes.PURCHASE_FAILED) &&
            this._purchasePollingCredentials &&
            !this._isPollingForSwapStatus) {
            // tslint:disable-next-line:no-floating-promises
            this._processPurchasePollingLoop(this._purchasePollingCredentials);
        }
    };
    RampInstantSDK.prototype._isConfiguredAsHosted = function () {
        return ['hosted-desktop', 'hosted-mobile'].includes(this._config.variant);
    };
    return RampInstantSDK;
}());
exports.RampInstantSDK = RampInstantSDK;
