// @ramp-network/ramp-instant-sdk/dist/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var EventSeverity;
(function (EventSeverity) {
    EventSeverity["VERBOSE"] = "VERBOSE";
    EventSeverity["DEBUG"] = "DEBUG";
    EventSeverity["INFO"] = "INFO";
    EventSeverity["WARNING"] = "WARNING";
    EventSeverity["ERROR"] = "ERROR";
    EventSeverity["CRITICAL"] = "CRITICAL";
})(EventSeverity = exports.EventSeverity || (exports.EventSeverity = {}));
var WidgetEventTypes;
(function (WidgetEventTypes) {
    WidgetEventTypes["WIDGET_CLOSE"] = "WIDGET_CLOSE";
    WidgetEventTypes["WIDGET_CONFIG_DONE"] = "WIDGET_CONFIG_DONE";
    WidgetEventTypes["WIDGET_CONFIG_FAILED"] = "WIDGET_CONFIG_FAILED";
    WidgetEventTypes["PURCHASE_CREATED"] = "PURCHASE_CREATED";
    WidgetEventTypes["PURCHASE_SUCCESSFUL"] = "PURCHASE_SUCCESSFUL";
    WidgetEventTypes["PURCHASE_FAILED"] = "PURCHASE_FAILED";
})(WidgetEventTypes = exports.WidgetEventTypes || (exports.WidgetEventTypes = {}));
var InternalEventTypes;
(function (InternalEventTypes) {
    InternalEventTypes["WIDGET_CLOSE_REQUEST"] = "WIDGET_CLOSE_REQUEST";
    InternalEventTypes["WIDGET_CLOSE_REQUEST_CANCELLED"] = "WIDGET_CLOSE_REQUEST_CANCELLED";
    InternalEventTypes["WIDGET_CLOSE_REQUEST_CONFIRMED"] = "WIDGET_CLOSE_REQUEST_CONFIRMED";
})(InternalEventTypes = exports.InternalEventTypes || (exports.InternalEventTypes = {}));
